### PR TITLE
[Candidate Parameters] Dynamic Family Tab (Redmine 11111)

### PR DESCRIPTION
**Pull request by stellarxo**
Dynamically add/remove family members

To Do: 
- [x] update candID list after addition of family member

### CODE REVIEW

File: modules/candidate_parameters/jsx/familyInfo.js [hunk 1]
```diff
@@ -1,353 +1,374 @@
...
+          that.setState(
+            {
+              Data: data,
+              isLoaded: true
+            }
+            );
+          var initialFamilyMembers = that.initializeFamilyMembers();
+          that.setState(
+            {
+              familyMembers: initialFamilyMembers
```
Comment (alisterdev) [inline modules/candidate_parameters/jsx/familyInfo.js]: Why not just do familyMembers: data.existingFamilyMembers (this looks over-engineered)

Comment (alisterdev) [inline modules/candidate_parameters/jsx/familyInfo.js]: Also, please put curlies on the same line as parenthesis, so eslint doesn't go crazy with indentation
```
this.setState({
 key : value
});
```

File: modules/candidate_parameters/jsx/familyInfo.js [hunk 2]
```diff
@@ -1,353 +1,374 @@
...
-                        );
-          },
-          error: function(data, errorCode, errorMsg) {
-            that.setState(
-              {
-                error: 'An error occurred when loading the form!'
+var FamilyInfo = React.createClass({
+  getInitialState: function() {
+    return {
+      relationshipOptions: {
```
Comment (alisterdev) [inline modules/candidate_parameters/jsx/familyInfo.js]: This doesn't need to be in the state, just declare as a variable inside render method

File: modules/candidate_parameters/jsx/familyInfo.js [hunk 3]
```diff
@@ -1,353 +1,374 @@
...
+        "1st_cousin": "First Cousin"
+      },
+      Data: [],
+      formData: {},
+      familyMembers: [],
+      updateResult: null,
+      errorMessage: null,
+      isLoaded: false,
+      loadedData: 0,
+      initialized: false
```
Comment (alisterdev) [inline modules/candidate_parameters/jsx/familyInfo.js]: `initialized` doesn't seem to be used anywhere

File: modules/candidate_parameters/jsx/familyInfo.js [hunk 4]
```diff
@@ -1,353 +1,374 @@
...
+    }
 
-      var relationshipRequired = false;
-      if (this.state.formData.FamilyCandID !== null &&
-                this.state.formData.FamilyCandID !== undefined
-            ) {
-        relationshipRequired = true;
-      }
+    var relationshipRequired = false;
+    if (this.state.formData.FamilyCandID !== null &&
```
Comment (alisterdev) [inline modules/candidate_parameters/jsx/familyInfo.js]: Can probably be written as 
```js
if (this.state.formData.FamilyCandID) {
}
```

File: modules/candidate_parameters/jsx/familyInfo.js [hunk 5]
```diff
@@ -1,353 +1,374 @@
...
+    candidateList[candID] = candID;
+
+    this.setState({
+      familyMembers: familyMembers
+    });
+
+    var myFormData = this.state.formData;
+    var self = this;
+    var formData = new FormData();
+    for (var key1 in myFormData) {
```
Comment (alisterdev) [inline modules/candidate_parameters/jsx/familyInfo.js]: You can use `let key` to have it in scope only inside the for loop

**Comment by stellarxo** [comment]
@alisterdev code changes pushed

**Comment by alisterdev** [comment]
@stellarxo All good! 👍

**Pull request merged**